server/src/main/java/io/crate/execution/dml/Indexer.java

```diff
@@ -747,38 +801,74 @@ public ParsedDocument index(IndexItem item) throws IOException {
-                continue;
-            }
-            ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(i);
-            if (value == null) {
-                continue;
+        if (indexOrder.isEmpty()) {
+            for (int i = 0; i < values.length; i++) {
+                Reference reference = columns.get(i);
+                Object value = valueForInsert(reference.valueType(), values[i]);
+                ColumnConstraint check = columnConstraints.get(reference.column());
+                if (check != null) {
+                    check.verify(value);
+                }
+                if (reference.granularity() == RowGranularity.PARTITION) {
+                    continue;
+                }
+                ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(i);
+                if (value == null) {
+                    continue;
+                }
+                translogWriter.writeFieldName(valueIndexer.storageIdentLeafName());
+                valueIndexer.indexValue(value, docBuilder);
             }
-            translogWriter.writeFieldName(valueIndexer.storageIdentLeafName());
-            valueIndexer.indexValue(value, docBuilder);
-        }
 
-        for (var entry : synthetics.entrySet()) {
-            ColumnIdent column = entry.getKey();
-            if (!column.isRoot()) {
-                continue;
+            for (var entry : synthetics.entrySet()) {
+                ColumnIdent column = entry.getKey();
+                if (!column.isRoot()) {
+                    continue;
+                }
+                Synthetic synthetic = entry.getValue();
+
+                Object value = synthetic.value();
+                if (value == null) {
+                    continue;
+                }
+                ValueIndexer<Object> indexer = synthetic.indexer();
+                translogWriter.writeFieldName(indexer.storageIdentLeafName());
+                indexer.indexValue(value, docBuilder);
             }
-            Synthetic synthetic = entry.getValue();
+        } else {
+            for (Reference ref : indexOrder) {
+                if (columns.contains(ref)) {
+                    int idx = columns.indexOf(ref);
+                    Object value = valueForInsert(ref.valueType(), values[idx]);
+                    ColumnConstraint check = columnConstraints.get(ref.column());
+                    if (check != null) {
+                        check.verify(value);
+                    }
+                    if (ref.granularity() == RowGranularity.PARTITION) {
+                        continue;
+                    }
+                    ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(idx);
+                    if (value != null) {
+                        translogWriter.writeFieldName(valueIndexer.storageIdentLeafName());
+                        valueIndexer.indexValue(value, docBuilder);
+                    }
```

Comment:
```suggestion
                    if (value != null) {
                        ValueIndexer<Object> valueIndexer = (ValueIndexer<Object>) valueIndexers.get(idx);
                        translogWriter.writeFieldName(valueIndexer.storageIdentLeafName());
                        valueIndexer.indexValue(value, docBuilder);
                    }
```

No need to retrieve the valueIndexer if it isn't used